### ISSUE TITLE
Build errors building phobos/druntime with ThinLTO on Linux

### ISSUE THREAD

**Issue by jondegenhardt** [ldc-developers/ldc#2348]
I'm trying to get the new support building phobos/druntime with LTO going. I'm getting errors. Likely I'm not setting something up correctly, hopefully you can help.

I'm trying to build linux builds on my Mac, so I'm using the dockerized ldc. Same as https://github.com/lindt/docker-ldc, except modified to use LDC 1.4 and add cmake to the installed apps. Ubuntu is 16.04, gcc is 5.4.0-6. I start the docker image with a bash shell, and can run ldc2 commands, these work fine, including compiling a hello-world test app.

First try: build runtime libraries with thin LTO:
```
$ ldc-build-runtime --dFlags="-flto=thin" BUILD_SHARED_LIBS=OFF
```
Then the test app:
```
ldc2 -release -flto=thin -Xcc=-fuse-ld=gold -L-L./ldc-build-runtime.tmp/lib test.d
```
This works fine, and the test app runs. However, adding in the -O switch gives errors:
```
# ldc2 -release -O -flto=thin -Xcc=-fuse-ld=gold -L-L./ldc-build-runtime.tmp/lib test.d

/tmp/lto-llvm-aabd89.o(.eh_frame+0xa093): error: undefined reference to '_d_eh_personality'
collect2: error: ld returned 1 exit status
Error: /usr/bin/gcc failed with status: 1
```

I also tried LTO full. A different error in that case:
```
$ ldc-build-runtime --reset --dFlags="-flto=full" BUILD_SHARED_LIBS=OFF
 => build succeeds

$ldc2 -flto=full -Xcc=-fuse-ld=gold -L-L./ldc-build-runtime.tmp/lib test.d

Global is external, but doesn't have external or weak linkage!
x86_fp80 (x86_fp80)* @_D3std4math4exp2FNaNbNiNeeZe
Global is external, but doesn't have external or weak linkage!
x86_fp80 (x86_fp80)* @_D3std4math5expm1FNaNbNiNeeZe
Global is external, but doesn't have external or weak linkage!
void (i8**, i8*)* @fiber_switchContext
Global is external, but doesn't have external or weak linkage!
i8* ()* @_D2rt19sections_elf_shared11getTLSRangeFNbNimmZ10endOfBlockFNaNbNiZPv
Global is external, but doesn't have external or weak linkage!
i8** ()* @_D4core7runtime19defaultTraceHandlerFPvZ16DefaultTraceInfo6__ctorMFZ10getBasePtrFNaNiZPPv
LLVM ERROR: Broken module found, compilation aborted!
collect2: error: ld returned 1 exit status
Error: /usr/bin/gcc failed with status: 1
```
The test app is very simple:
```
void main (string[] args)
{
    import std.stdio;

    writeln("Hello World");
}
```
Anything I should be doing differently?

Update: Some version info
```
$ ldc2 --version
LDC - the LLVM D compiler (1.4.0):
  based on DMD v2.074.1 and LLVM 4.0.1
  built with LDC - the LLVM D compiler (1.4.0)
  Default target: x86_64-unknown-linux-gnu
  Host CPU: haswell
  http://dlang.org - http://wiki.dlang.org/LDC
```

**Comment by klickverbot** [comment]
> Anything I should be doing differently?

I haven't seen that particular set of errors before at all, but it seems like it might be due to an LLVM bug or version mismatch. You could try building against LLVM 5 instead, double-checking that there is only one version around.

**Comment by jondegenhardt** [comment]
Do you mean build ldc from source with LLVM 5  rather than use the prebuilt release package?

**Comment by klickverbot** [comment]
Oh, sorry, I didn't realise the image was using the prebuilt binaries.

**Comment by jondegenhardt** [comment]
Could it be the ubuntu version or gcc version? I was trying to stick with the docker-ldc in the Docker Store (https://store.docker.com/community/images/dlanguage/ldc). It pulls the pre-built ldc packages. It's still using ldc 1.4.0-beta1. It's based on ubuntu 16.04. I switched it to ldc 1.4.0, but I left the ubuntu version. The gcc version is 5.0.4-6.

It's easy enough to change any of these things. Perhaps if you can describe the environment used to test the prebuilt binaries I can try to match it.

**Comment by kinke** [comment]
I can reproduce the issue with ThinLTO libs and `-O` with 1.4 final on Ubuntu 16.04.

**Comment by kinke** [comment]
ThinLTO and `-O` is working with LLVM 5.0 for me though [using LDC branch `merge-2.075`, but that shouldn't matter].

**Comment by jondegenhardt** [comment]
> hello-world at least works with the FullLTO libs, with both -flto=full|thin and with/without -O.
With LLVM 5.0?

**Comment by kinke** [comment]
Yep.

**Comment by kinke** [comment]
[If you're going to rebuild against LLVM 5.0, either use latest master or `-DLDC_WITH_LLD=OFF` in the CMake command line.]

**Comment by jondegenhardt** [comment]
I'm not going to rebuild against LLVM 5.0 just yet (rather busy with work tasks).

My goal is to see if I can get a travis-ci build going with phobos/druntime LTO turned on. If I can get it working, I can release the eBay tsv utilities this way. But, I don't want to do this unless it's working with a prepackaged ldc build. Mostly, if anyone clones the repo, I'd like them to be able to try building it as well. Needing to rebuild LDC is a bit much.

By itself, turning on LTO for the tsv utilities is a rather low priority. However, given the performance gains seen while investigating #2168, I've had some thoughts about making a short blog post 
 (@JohanEngelen's suggestion). I'd like to get it released first though.

**Comment by JohanEngelen** [comment]
@jondegenhardt Thanks a lot for testing LTO!

@kinke I've seen the same assert with ThinLTO. Need to find the time to minimize and report to LLVM...

**Comment by jondegenhardt** [comment]
Same result with Ubuntu 14.04.

**Comment by jondegenhardt** [comment]
Some follow-up: The issue with Linux ThinLTO with the `-O` persists with the 1.5.0 release. ie.
```
$ ldc2 --version
LDC - the LLVM D compiler (1.5.0):
  based on DMD v2.075.1 and LLVM 5.0.0
  built with LDC - the LLVM D compiler (1.5.0)
  Default target: x86_64-unknown-linux-gnu
  Host CPU: haswell
  http://dlang.org - http://wiki.dlang.org/LDC

$  cat helloworld.d
void main(string[] args)
{
    import std.stdio;
    writeln("Hello World!");
}

$  ldc-build-runtime --dFlags="-flto=thin" BUILD_SHARED_LIBS=OFF
Creating build directory: ldc-build-runtime.tmp
Downloading LDC source archive: https://github.com/ldc-developers/ldc/releases/download/v1.5.0/ldc-1.5.0-src.zip
Invoking: ["cmake", "-DLDC_EXE_FULL=/dlang/ldc-1.5.0/bin/ldc2", "-DD_VERSION=2", "-DDMDFE_MINOR_VERSION=0", "-DDMDFE_PATCH_VERSION=75", "-DLDC_TARGET_PRESET=", "-DTARGET_SYSTEM=AUTO", "-DD_FLAGS=-flto=thin", "-DRT_CFLAGS=", "-DLD_FLAGS=", "-DBUILD_SHARED_LIBS=OFF", "/src/ldc-build-runtime.tmp/ldc-src/runtime"]

...

[100%] Built target phobos2-ldc
Runtime libraries built successfully into: /src/ldc-build-runtime.tmp

$  ldc2 -O -flto=thin -linker=gold -L-L./ldc-build-runtime.tmp/lib helloworld.d
/tmp/lto-llvm-f5a8b2.o(.eh_frame+0x9d53): error: undefined reference to '_d_eh_personality'
collect2: error: ld returned 1 exit status
Error: /usr/bin/gcc failed with status: 1
```

Compilation succeeds if the `-O` switch is removed. Using Full LTO also succeeds.
The above was run on Ubuntu 16.04.

**Comment by jondegenhardt** [comment]
Tried Linux ThinLTO with LDC 1.7.0 (Ubuntu 16.04), which uses LLVM 5.0.1. Didn't try the `helloworld.d` program listed in the previous comment, just used my tsv-utils build. But, still got the same build failure: 
```
/tmp/lto-llvm-a686ba.o(.eh_frame+0xabf3): error: undefined reference to '_d_eh_personality'
```

**Comment by JohanEngelen** [comment]
@jondegenhardt At Weka, I was able to get their huge codebase built with ThinLTO with LDC 1.6.0, LLVM 5.0.1, on Ubuntu 14.04, ld.gold (GNU gold (GNU Binutils 2.26.51.20160526) 1.11) .  What am I doing wrong? ;-)

Could you try to minimize the issue in terms of LLVM files? (perhaps start with a copy of phobos/druntime with pretty much everything removed)

**Comment by jondegenhardt** [comment]
@JohanEngelen Maybe I'm misunderstanding the idea, but it sounds like a fair bit of work to strip druntime/phobos down to smallest set that would actually work. Right now I don't have time to make that kind of investment.

However, I did try the `helloworld.d` sample again, which is pretty minimal, and it still fails. Perhaps a start is to see if you can reproduce that.

Because I'm on MacOS, I use Docker to do Linux builds. When I originally reported this I did the pull of the pre-built binary package myself. This time I used the ldc docker image available from DockerHub (which is what I use to build on Travis CI.) The run log is attached:  [log.txt](https://github.com/ldc-developers/ldc/files/1612893/log.txt)

The key parts:
```
$ cat helloworld.d
void main(string[] args)
{
    import std.stdio;
    writeln("Hello World!");
}
$ docker run --rm -ti -v $(pwd):/src dlanguage/ldc:1.7.0 'ldc-build-runtime --dFlags="-flto=thin" BUILD_SHARED_LIBS=OFF && ldc2 -release -O -flto=thin -Xcc=-fuse-ld=gold -L-L./ldc-build-runtime.tmp/lib helloworld.d'
Creating build directory: ldc-build-runtime.tmp
Downloading LDC source archive: https://github.com/ldc-developers/ldc/releases/download/v1.7.0/ldc-1.7.0-src.zip
Invoking: ["cmake", "-DLDC_EXE_FULL=/dlang/ldc-1.7.0/bin/ldc2", "-DD_VERSION=2", "-DDMDFE_MINOR_VERSION=0", "-DDMDFE_PATCH_VERSION=77", "-DLDC_TARGET_PRESET=", "-DTARGET_SYSTEM=AUTO", "-DD_FLAGS=-flto=thin", "-DRT_CFLAGS=", "-DLD_FLAGS=", "-DBUILD_SHARED_LIBS=OFF", "/src/ldc-build-runtime.tmp/ldc-src/runtime"]
-- The C compiler identification is GNU 5.4.0

   [snip]

Runtime libraries built successfully into: /src/ldc-build-runtime.tmp
/tmp/lto-llvm-5964dd.o(.eh_frame+0xa1eb): error: undefined reference to '_d_eh_personality'
collect2: error: ld returned 1 exit status
Error: /usr/bin/gcc failed with status: 1
```

Note that the `docker run` command includes both the `ldc-build-runtime` call and the `ldc2` call to build `helloworld`. The error at the end is for the `ldc2` call. (Note also the single quotes around the command passed to docker.)

**Comment by JohanEngelen** [comment]
OK cool, thanks for all the steps
(starved on time myself too)

**Comment by kinke** [comment]
Jon, have you ever tried the combo FullLTO libs + ThinLTO user code?

**Comment by jondegenhardt** [comment]
No, haven't tried mixing Full and Thin like that. My understanding was that the two were incompatible. I can certainly try it if it would be helpful.

**Comment by kinke** [comment]
It worked in my earlier tests, so maybe worth giving a shot.

**Comment by jondegenhardt** [comment]
Okay, I'll give it a try. It will likely be a few hours before I can get to it.
Using FullLTO on both dlibs and user code is fine.

**Comment by JohanEngelen** [comment]
Ahh wait... I forgot. I didn't build ThinLTO druntime/Phobos... Disregard.

**Comment by jondegenhardt** [comment]
Modified the title to reflect the narrower scope of the issue. FullLTO with phobos/druntime on Linux now works correctly.

**Comment by jondegenhardt** [comment]
The helloworld app builds and runs with dlibs built FullLTO and app code built ThinLTO. Haven't tried anything else.

**Comment by kinke** [comment]
@jondegenhardt: Is this still an issue? We've been shipping ThinLTO default libs on Linux for some time now (built on Ubuntu 14.04 with backported gcc 6).

**Comment by jondegenhardt** [comment]
@kinke Far as I know it's still an issue. Last version I checked was LDC 1.7, it was still an issue then. I'll add it to my todo list to check the current release and put a status update here.

**Comment by jondegenhardt** [comment]
@kinke Likely fixed, though I saw some other errors while trying to reproduce them. I expect they are different. I tested with the prebuilt release binaries for LDC 1.11.0-beta-2. Specifics...

I can no longer reproduce this on the simple apps (helloworld.d). For `tsv-utils` all the tools except one build fine, that is `tsv-filter` (which causes other issues reported about `tsv-utils`, apparently due to the long argument list passed to `getopt`).

For these I'm using `ldc-build-runtime` to create the ThinLTO version of druntime/phobos. However, if I switch to the newer method based on the shipped phobos/druntime then `tsv-filter` is fine. However, `tsv-filter` fails with LTO=full, which is new.

I'm inclined to close this issue and open the `tsv-filter` cases as a separate issue.
